### PR TITLE
create orchestrator peer dbus on connection request

### LINKED ISSUE #12: Move from POC: [Orchestrator] create peer bus
## Please describe what you would like to see

On a connection request from a node, the orchestrator creates a new peer dbus. 

## Please describe the solution you'd like

For reference, see [the implementation from the POC](https://github.com/alexlarsson/orch/blob/main/orch.c#L500-L562). 
Maybe use a factory pattern to create peer dbus (e.g. see [here](https://chuan-zhang.medium.com/implementing-factory-pattern-in-c-c6f9b02bd226) for factory pattern in C)

## Please describe your use case

Creating a peer dbus that does not use the dbus daemon. 

image

### PR DESCRIPTION
**Pull request by engelmi**
This PR fixes #12 by
- implementing a general peer dbus as well as a two adaptions for a server-side and client-side peer dbus
- adding a `PeerManager` to the `orchestrator` that contains and manages the server-side dbus
- creating a server-side peer dbus on a connection request, linking it with the client-side one

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -7,7 +7,6 @@ project(
     'c_std=gnu17',     # Adds "-std=gnu17".  Includes GNU 17 extensions.
     'warning_level=2', # Adds "-Wextra".  Enables additional warnings. 
     'debug=true',      # Adds "-g".  Object files include debugging symbols.
-    'werror=true'      # Adds "-Werror".  Treat warnings as errors.
```
Comment (sdunnagan) [inline meson.build]: We should continue treating warnings as errors.

Comment (engelmi) [reply meson.build]: Good catch. Didn't intend to remove it.

File: libchihua-msg/src/node.c [hunk 2]
```diff
@@ -15,51 +15,61 @@ Node *node_new(const NodeParams *params) {
...
+        if ((*node)->systemd_dbus != NULL) {
+                fprintf(stdout, "Freeing allocated dbus to local systemd dbus...\n");
+                sd_bus_unrefp(&(*node)->systemd_dbus);
         }
 
         free(*node);
 }
 
-bool node_start(const Node *node) {
+bool node_start(Node *node) {
```
Comment (sdunnagan) [inline libchihua-msg/src/node.c:66]: The naming seems confusing.  What does it mean to start a node?  What we're really doing is starting an event loop.  Maybe change this to ```node_start_event_loop()```?

File: libchihua-msg/include/orchestrator/connection-server.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "../common/list.h"
+#include "../common/memory.h"
+#include "peer-manager.h"
+
+typedef struct {
+        sd_event_source *sd_event_source;
+
+        PeerManager *peer_manager;
+} ConnectionServer;
```
Comment (sdunnagan) [inline libchihua-msg/include/orchestrator/connection-server.h]: I am learning D-Bus, so I'm not too opinionated yet about our use of it.  But I find the name of this type confusing.  Does it serve up D-Bus connections?  And to whom would it serve them?

**Review by sdunnagan — APPROVED** [review]

File: libchihua-msg/include/orchestrator/connection-server.h (hunk 3 above)
Comment (mkemel) [reply libchihua-msg/include/orchestrator/connection-server.h]: I agree. I think something like ConnectionManager would be better.

**Review by mkemel — APPROVED** [review]

File: libchihua-msg/src/node.c (hunk 2 above)
Comment (engelmi) [reply libchihua-msg/src/node.c:66]: Yes, it simply starts the event loop (at least right now). However, before starting the event loop no processing will happen. For example, even though we create and start the peer dbus in `node_new`, it doesn't do anything else till the event loop is started. So starting the event loop means to start all the functions of a node - its essentially the core of a node. Thats why I'd keep the name as it is for now. 

But I accidentally removed the `const` - I'll re-add it.

File: libchihua-msg/include/orchestrator/connection-server.h (hunk 3 above)
Comment (engelmi) [reply libchihua-msg/include/orchestrator/connection-server.h]: Initially, I had two objectives in mind:
- separating the creation of the TCP socket and accepting incoming connections from creating the server-side peer dbus (which is done by the `PeerManager`)
- having a struct to control the live-cycle of the `accept_fd`, `event_source` and `PeerManager`

However, looking at it now, it seems that `ConnectionServer` struct is not really needed. The library takes care of closing `accept_fd` by calling `sd_event_source_set_io_fd_own(event_source, true)` and the `PeerManager` is passed in as parameter. I'll have another look and probably remove the struct.

**Pull request merged**